### ISSUE TITLE
When using nuclei as module/lib subsequently scans will not return any result

### ISSUE THREAD

**Issue by krijohs** [projectdiscovery/nuclei#2682]
### Nuclei version:
2.7.7

### Current Behavior:
I'm using nuclei as a module/library in a Go project, and want to scan more than one time.
After the first scan any subsequent scan will not return any result.

### Expected Behavior:
That nuclei module can be used to run more than one scan.

### Steps To Reproduce:
```go
package main

import (
	"fmt"
	"log"
	"os"
	"path"

	"github.com/logrusorgru/aurora"
	"github.com/projectdiscovery/nuclei/v2/pkg/catalog/config"
	"github.com/projectdiscovery/nuclei/v2/pkg/catalog/disk"
	"github.com/projectdiscovery/nuclei/v2/pkg/catalog/loader"
	"github.com/projectdiscovery/nuclei/v2/pkg/core"
	"github.com/projectdiscovery/nuclei/v2/pkg/core/inputs"
	"github.com/projectdiscovery/nuclei/v2/pkg/output"
	"github.com/projectdiscovery/nuclei/v2/pkg/parsers"
	"github.com/projectdiscovery/nuclei/v2/pkg/protocols"
	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/hosterrorscache"
	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/interactsh"
	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/protocolinit"
	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/protocolstate"
	"github.com/projectdiscovery/nuclei/v2/pkg/testutils"
	"github.com/projectdiscovery/nuclei/v2/pkg/types"
	"go.uber.org/ratelimit"
)

func main() {
	res1, err := nuclei("host2.com")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("length of res1: %d\n", len(res1))

	res2, err := nuclei("host2.com")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("length of res2: %d\n", len(res2))

	res3, err := nuclei("host1.com")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("length of res3: %d\n", len(res3))
}

func nuclei(host string) ([]output.ResultEvent, error) {
	cache := hosterrorscache.New(30, hosterrorscache.DefaultMaxHostsCount)
	defer cache.Close()

	mockProgress := &testutils.MockProgressClient{}

	result := make([]output.ResultEvent, 0)
	outputWriter := testutils.NewMockOutputWriter()
	outputWriter.WriteCallback = func(event *output.ResultEvent) {
		result = append(result, *event)
	}

	defaultOpts := types.DefaultOptions()
	protocolstate.Init(defaultOpts)
	protocolinit.Init(defaultOpts)

	defaultOpts.ExcludeTags = config.ReadIgnoreFile().Tags

	interactOpts := interactsh.NewDefaultOptions(outputWriter, nil, mockProgress)
	interactClient, err := interactsh.New(interactOpts)
	if err != nil {
		return nil, err
	}
	defer interactClient.Close()

	home, err := os.UserHomeDir()
	if err != nil {
		return nil, err
	}

	catalog := disk.NewCatalog(path.Join(home, "nuclei-templates"))
	executerOpts := protocols.ExecuterOptions{
		Output:          outputWriter,
		Options:         defaultOpts,
		Progress:        mockProgress,
		Catalog:         catalog,
		RateLimiter:     ratelimit.New(150),
		Interactsh:      interactClient,
		HostErrorsCache: cache,
		Colorizer:       aurora.NewAurora(true),
		ResumeCfg:       types.NewResumeCfg(),
	}
	engine := core.New(defaultOpts)
	engine.SetExecuterOptions(executerOpts)

	workflowLoader, err := parsers.NewLoader(&executerOpts)
	if err != nil {
		return nil, err
	}

	executerOpts.WorkflowLoader = workflowLoader

	configObject, err := config.ReadConfiguration()
	if err != nil {
		return nil, err
	}

	store, err := loader.New(loader.NewConfig(defaultOpts, configObject, catalog, executerOpts))
	if err != nil {
		return nil, err
	}
	store.Load()

	input := &inputs.SimpleInputProvider{Inputs: []string{host}}
	_ = engine.Execute(store.Templates(), input)
	engine.WorkPool().Wait()
	return result, nil
}
```

**Comment by Mzack9999** [comment]
I'm unable to reproduce the behavior:
```console
$ go run .
[INF] Using Interactsh Server: oast.online
length of res1: 19
length of res2: 19
length of res3: 19
$ go run .
[INF] Using Interactsh Server: oast.online
length of res1: 19
length of res2: 19
length of res3: 19
$ cat main.go
...
func main() {
	target := "http://127.0.0.1:8000"
	res1, err := nuclei(target)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("length of res1: %d\n", len(res1))

	res2, err := nuclei(target)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("length of res2: %d\n", len(res2))

	res3, err := nuclei(target)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("length of res3: %d\n", len(res3))
}
...
```
Have you tried prefixing the target with `http` or `https`?

**Comment by krijohs** [comment]
Thats wierd, For me it doesn't matter if using a target with `http` or `http` , and the result after the first scan is always 0, sometimes 1 even if the same target/host is used for the scans. I've also ran the scans from entirely different host and network.

When I set a breakpoint to check what `result` slice consists of and whats being being added to it after the first scan, I can see that it includes results from the previous scan and the returned result differs from when running without setting breakpoints. So for me it seems like there is a race condition somewhere in the code but I might be wrong.

**Comment by Mzack9999** [comment]
@krijohs apologies for the late reply - I've been trying this multiple times against our test domain `scanme.sh`. I confirmed a few times I got 21 vulnerabilities and other times 19, primarily due to network instability or different DNS resolvers picked up. The discrepancies you are facing are probably not due to the tool but network conditions. Not sure if increasing retries/timeouts or reducing parallelism/concurrency might help.

**Comment by krijohs** [comment]
No problem.
The test I linked to is running in a docker container so internal network is only used except for the external calls nuclei does. So not sure if network is the issue here.
I've updated to latest version 2.8.1 but it discovers 25 on first call then 3,17 or 20. 
When running the scan using the cli the same amount of discovered vulnerabilities are returned on subsequent calls, so this also makes me believe the network is not the issue here.
